libsolidity/analysis/TypeChecker.cpp

```diff
@@ -607,6 +607,12 @@ bool TypeChecker::visit(StructDefinition const& _struct)
 				auto const& typeName = dynamic_cast<UserDefinedTypeName const&>(*member->typeName());
 				check(&dynamic_cast<StructDefinition const&>(*typeName.annotation().referencedDeclaration), parents);
 			}
+			else if (auto arrayType = dynamic_cast<ArrayType const*>(type(*member).get()))
```

Comment:
Oh this is very interesting! We have some kind of duplication in `StructType::recursive()` which also disallows dynamic arrays because the purpose is different. Because of that I guess it is still better to keep both versions, but perhaps this implementation here could be simplified to use `CycleDetector` class?

Reply:
Looking a bit closer at ``StructType::recursive`` I think I forgot to take care about multi-dimensional arrays as well :). But using ``CycleDetector`` makes sense in any case. I'll update and add tests for the multi-dimensional array case.

Reply:
There is no such thing as multi-dimensional arrays. Are you talking about dynamic arrays? In contrast to `::recursive`, they are fine at this point because here, the issue is that we want to check whether the storage size of the struct is finite, which is the case for dynamically-sized arrays that recursively contain the struct itself.

Reply:
To me ``struct S { S[1][1] s; };`` is a multi-dimensional array and this is bad as well - dynamic arrays are fine as in ``struct S { S[][1] s; };``, ``struct S { S[1][] s; };``, ``struct S { S[][] s; };`` as well as ``struct S { S[] s; };`` are all fine.

Reply:
"Arrays of arrays" is what I meant.